### DISCUSSION TITLE
Deep Linking Support

### DISCUSSION THREAD

**Discussion by DavidKmn** [pointfreeco/swift-composable-architecture#1006] [Q&A] [answered]
Recently watched the navigation series from Point Free and was wondering how would deep linking be approached in the composable architecture? Had a look and did not find any examples, also did not see deep linking capabilities in the isowords app...

**Comment by mbrandonw** [comment, marked as answer]
Deep linking in TCA would happen in much the same way as it did in the last episode of the navigation series ([this one](https://www.pointfree.co/collections/swiftui/navigation/ep168-swiftui-navigation-the-point)). You first need to figure out how you are going to process the incoming deep link URL in order to figure out where the user needs to be routed (we used our parsing library to accomplish this, but you can use anything) and then you construct the app state to represent where you want to navigate to.

Although isowords doesn't need deep linking it does need to route push notifications, which is similar. We parse an incoming push notification payload and figure out where to navigate the user here:

https://github.com/pointfreeco/isowords/blob/b87f4c61c0523564adf17c031f9fa4b88a40d631/Sources/AppFeature/AppView.swift#L235-L262

In particular, if the push notification is the daily challenge report we navigate the user to the daily challenge screen here:

https://github.com/pointfreeco/isowords/blob/b87f4c61c0523564adf17c031f9fa4b88a40d631/Sources/AppFeature/AppView.swift#L257-L260

Hope that helps a little bit.

**Reply by DavidKmn** [reply]
Hey Brandon, thanks for this, have looked at this yes. Loved the Parsing lib updates you guys did, using that already, great addition of Result Builders and errors. 

So here are my thoughts. First off I understand the idea of setting the state in this way, however what if we want deep linking workflow where each step can be an async operation with cross dependencies. If you are familiar with RIBs they have a concept of a Workflow and an ActionableItem. Here is a good outlay https://github.com/uber/RIBs/wiki/iOS-Tutorial-4. Wondering if something like this is possible using TCA.

Second more important concert I have that flows out is the tight coupling of feature modules. Views in each feature know about other features views, meaning they import other feature modules, which I don't thing is great. This is a solved issue in RIBs UIKit where a correct implementation of the architecture would allow for complete feature module decoupling. However, SwiftUI makes this inherently difficult to achieve since routing to other views is tightly coupled (unless we just resort to wrapping SwiftUI in UIKit and routing using standard UIKit Navigation primitives), views route to each other and know about each other, I don't see how TCA can be used when we want no feature to feature dependencies. Have you thought about approaches to remove feature to feature dependencies, RIBs solves this by using module construction delegation up to some other composition root who does actually import (in most cases this is in the main app target which is fine).

**Reply by johnpatrickmorgan** [reply]
Hi David, I’m a bit late to this thread but I wanted to direct you towards https://github.com/johnpatrickmorgan/TCACoordinators. It allows you to write screen features that are entirely independent of one another and unaware of their place in a navigation flow.

**Reply by DavidKmn** [reply]
Thanks @johnpatrickmorgan, just had a look. Not sure how this would work for this situation. Let's say I have an app with 2 Feature modules, Home and Products, these are separate SMP modules. I want to embed logic to allow Home to open the Products screen on a button tap. We do not want a cross feature dependency therefore we cannot directly just open Product from home. Meaning the only way is to allow some higher level module to control the navigation flow. So from my understanding you are suggesting that we have an AppCoordinator and 2 FeatureCoordinators that live in each module, the question is how does HomeFeatureCoordinator communicate and use the ProductFeatureCoordinator?

**Reply by johnpatrickmorgan** [reply]
Thanks @DavidKmn - assuming `Home` and `Product` are single-screen features then they wouldn't need their own coordinators. They would each exist in their own modules with no knowledge of each other. The action that triggers navigation from `Home` to `Product` would go unimplemented in the Home module. A third module would import both Home and Product and compose them within a Coordinator feature using `TCARouter`, which would respond to the trigger action by showing the Product screen.  Similar to how the individual `Home`, `NumbersList` and `NumberDetail` features are independent of each other (and could be separated into their own modules) in this example:

https://github.com/johnpatrickmorgan/TCACoordinators/blob/b902f399295206d100a806313d1e79a4a3533e34/TCACoordinatorsExample/TCACoordinatorsExample/Screen.swift

If `Home` were a multi-screen feature with its own coordinator, it wouldn't work so easily. 

Apologies if I've misinterpreted your issue!

**Reply by DavidKmn** [reply]
@johnpatrickmorgan I see what you mean, yeah a solution for sure. I actually went ahead with a custom architecture instead of TCA, was wondering where would you suggest the management of the routes happens. Say we take a basic MVVM architecture, we now go back to a MVVM-C architecture where the View sends actions to the VM and the VM then updates routes in the Coordinator?

**Reply by johnpatrickmorgan** [reply]
Hi @DavidKmn, TCACoordinators is built on top of [FlowStacks](https://github.com/johnpatrickmorgan/FlowStacks), which can be used in non-TCA projects. There's an [example in the README](https://github.com/johnpatrickmorgan/FlowStacks#using-view-models) of how it can be used with ViewModels. The coordinator is itself a View with its own ViewModel that holds onto an array of Routes. Hope that helps!

**Reply by DavidKmn** [reply]
Thanks for this @johnpatrickmorgan, sorry for the late reply, somehow missed the Github notification. Had a look at the example, I was wondering why not keep the routes in the view model, any reason why there is a need for a separate view model for the coordinator?

**Reply by johnpatrickmorgan** [reply]
> I was wondering why not keep the routes in the view model, any reason why there is a need for a separate view model for the coordinator?

Thanks @DavidKmn. Did you mean to ask 'why not keep the routes in the view'? If so, yes, you could certainly avoid having a view model for the coordinator and just have the coordinator view hold on to the routes itself. Though if you're using view models for individual screen views, I don't see why you wouldn't take the same approach for the coordinator view too. 

In any case, both approaches should work well.
